### ISSUE TITLE
Render image on nodejs server

### ISSUE THREAD

**Issue by NitroCipher** [bs-community/skinview3d#52]
I am working on a project, and would like to be able to send rendered images from this to clients.

I have tried to modify it to use three-software-renderer, rather than WebGLRenderer, and I cant see to get that to work.

Do you have any example code of this project being used server-side?

**Comment by g-plane** [comment]
Hello, this library is only for client-side used.

**Comment by Hacksore** [comment]
You could definitely use something like [puppeteer](https://github.com/puppeteer/puppeteer) to render via a browser on the server-side and capture images.

**Comment by skerit** [comment]
It's actually not that hard to make this work on the server-side.
I had to modify the bundle a bit, and added this to the top:
(this basically shims a few objects skinview3d expects to be there)

```javascript
var document;

if (is_browser) {
	document = window.document;
} else {
	const NodeCanvas = require('node-canvas-webgl');

	document = {
		createElement : function(element) {

			if (element == 'canvas') {
				return NodeCanvas.createCanvas();
			} else if (element == 'img') {
				return new NodeCanvas.Image();
			}

		}
	};

	eval('var window = {}');
	eval('var HTMLImageElement = null');
	eval('var HTMLVideoElement = null');
	eval('var HTMLCanvasElement = null');
	eval('var ImageBitmap = null');
	eval('var OffscreenCanvas = null');

	HTMLImageElement = NodeCanvas.Image;

	HTMLVideoElement = ImageBitmap = OffscreenCanvas = function() {};
	HTMLCanvasElement = function() {};

	window.devicePixelRatio = 1;
	window.requestAnimationFrame = function(fnc) {
		fnc();
	};
}
```

And to actually render something, I did this:

```javascript
let createCanvas = require('node-canvas-webgl').createCanvas;
const canvas = createCanvas(200, 250);
const Skinview3D = require('local/path/to/skinview3d.bundle.js');

let skinview = new Skinview3D.SkinViewer({
	canvas : canvas,
	width  : 200,
	height : 250,
	renderPaused : true
});

await skinview.loadSkin('https://full.url/to/skin');

skinview.render();

let buffer = canvas.toBuffer();
```

**Comment by Hacksore** [comment]
@skerit Would you be up for adding a PR to add support for this natively?

**Comment by g-plane** [comment]
I'll investigate this solution.

**Comment by g-plane** [comment]
I've failed to do it:
- Three.js didn't declare it as an ES Module by setting `type` as `module` in `package.json`.
- `skinview3d` assumes we're running code in browser, so we have to inject polyfill with `jsdom`, which isn't easy and bring boilerplate code.
- Currently we have to hack `document.createElement` like this:
```js
const originalFn = document.createElement;
document.createElement = (/** @param {string} */ tag) => {
	switch (tag) {
		case "canvas":
			return canvas.createCanvas(0, 0);
		case "img":
			return new canvas.Image();
		default:
			return originalFn(tag);
	}
};
```
- Normally Node.js will read a file as buffer if we don't specify encoding, but `skinview3d` doesn't support buffer. However, Node.js supports read a file as a Base64 URL. I haven't tested it yet.
- I've encountered an error: `_canvas.addEventListener is not a function`.
